### PR TITLE
REST data Panache reorganisation

### PR DESCRIPTION
**Pull request by gytis**
https://github.com/quarkusio/quarkus/issues/9293
https://github.com/quarkusio/quarkus/issues/9261

Sorry for a big PR. I wanted to squeeze the main changes in before the Tuesday's release. I kept two separate commits, but I can squash them if you prefer it.

The main changes are:
* Rename the extension and all related modules from `quarkus-panache-rest-*` to `quarkus-*-rest-data-panache`
* Move URL building logic (`Path` annotations and `Location` headers etc.) to one place outside of the method implementors.
* Build controller path based on a resource rather than entity.
* Make path configurable
* Split `@PanacheRestResource` configuration annotation to two `@ResourceProperties` (to be used on an interface level and `@OperationProperties` (to be used on a method).

### CODE REVIEW

**Comment by gytis** [comment]
cc @geoand

**Comment by geoand** [comment]
Thanks, I'll take a look first thing Monday.

I assume you also want a review from @FroMage ?

**Comment by gytis** [comment]
Yeah that would be helpful.

**Comment by geoand** [comment]
Can you please rebase onto master to pick up a CI fix?

Thanks

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (geoand) [inline extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:26]: I assume this is used in addition to whatever the Class uses, right?

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java
Comment (geoand) [inline extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java:7]: Any reason why this was renamed?

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/properties/ResourcePropertiesAccessor.java
Comment (geoand) [inline extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/properties/ResourcePropertiesAccessor.java:41]: Minor nipick (can be fixed later): The use of `controller` seems a little out of place. Perhaps use `resource` instead?

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResourceName.java
Comment (geoand) [inline extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResourceName.java:12]: This seems to be unused now, no?

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (geoand) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:26]: Looking at the tests, I see it does :)

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (geoand) [inline extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:30]: Not that I think this will ever be an issue, but by using `""` as the default value, you are effectively preventing users from making an panache rest data resource without a path. Now that's probably A good thing, but I am just bringing it up in case you hadn't considered it. In any case it would probably be safer to use some nonsensical default valur

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (gytis) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:26]: Yes, I've added a `@Path` annotation to the class too

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java
Comment (gytis) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java:7]: Before it was already used by a processor which would extract separate bits of information and only pass what's needed to a specific method implementor. I made the method implementors more abstract now so processor wouldn't have to know too much details about them. Now it just passes this object to each implementor. So I assumed the name without "BuildItem" suffix would be more suitable. What do you think?

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/properties/ResourcePropertiesAccessor.java
Comment (gytis) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/properties/ResourcePropertiesAccessor.java:41]: True. I'll take a note of this and will fix it separately after the release.

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (gytis) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:30]: Hmm, we could maybe change this later too. Adding a nonsense value and based on that not add an annotation at all. I need to add some default value because otherwise user will always have to specify a path when using this annotation.

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java
Comment (geoand) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java:7]: It just feels weird to have being a build item and not being called that. What we usually do in similar cases is to have a build item that just holds a reference to the regular class. Then in the non-build-step code you can just pass around the regular class. The build step classes would of course the build items to communicate.

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (geoand) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:30]: Yeah, I understand the need for a default value, just saying that `""` is not the proper one :)

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResourceName.java
Comment (gytis) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResourceName.java:12]: True, I had a use for it initially. I'll get rid of it.

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java
Comment (gytis) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java:7]: Yeah that makes sense, I'll extract it to a separate class

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (gytis) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:30]: Shall we change it separately after the release? I don't want to hold this PR for too long.

Comment (geoand) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:30]: Sure yeah

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResourceName.java
Comment (gytis) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/ResourceName.java]: Removed

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/properties/ResourcePropertiesAccessor.java
Comment (gytis) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/properties/ResourcePropertiesAccessor.java:41]: Renamed

File: extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java
Comment (gytis) [reply extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/PanacheCrudResourceInfo.java]: Done

**Review by geoand — APPROVED** [review]
Seems fine to me!

**Comment by gytis** [comment]
@geoand I've added modifications as a separate commit. I have to either keep them like this or squash all three commits (including the rename). Otherwise squashing first and third commits causes a lot of conflicts, so will take a lot of time to go through them.

**Comment by geoand** [comment]
I think we can live with the extra commit :)

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (FroMage) [inline extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:12]: So you were against CRUD in the module name, but not here in the resource name? Should it be `PanacheEntityDataResource`?

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (FroMage) [inline extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:14]: Given that you've called the class annotation `ResourceProperties` after JAX-RS's (horrible) choice, is there a colloquial term for methods too or did you pick `Operation` at random?

FTR JAX-RS calls methods "resource methods":

> Resource methods are methods of a resource class annotated with a request method designator. They are
used to handle requests and MUST conform to certain restrictions described in this section.
A request method designator is a runtime annotation that is annotated with the @HttpMethod annotation.
JAX-RS defines a set of request method designators for the common HTTP methods: @GET, @POST, @PUT,
@DELETE, @PATCH, @HEAD and @OPTIONS. Users may define their own custom request method designators
including alternate designators for the common HTTP methods

I just want to make sure people are not going to object at us calling this an `Operation`.

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/OperationProperties.java
Comment (FroMage) [inline extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/OperationProperties.java:13]: Could probably do with javadoc to mention this annotation is not mandatory unless you want to override the defaults.

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (FroMage) [inline extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:13]: Could probably do with javadoc to mention this annotation is not mandatory unless you want to override the defaults.

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (FroMage) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:26]: Probably worth mentioning in the javadoc both here and on the `OperationsProperties.path` one to explain they accumulate.

**Review by FroMage — CHANGES_REQUESTED** [review]
Thanks for the quick work, and sorry about the late review. Aren't we missing the docs too?

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:14]: I called them operations, because that term seems to be used in W3C https://www.w3.org/2001/sw/wiki/REST#Operations_on_Elements. I'm not a purist of any terminology. So whichever is more acceptable I'm fine with it.

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:12]: It does sounds better to me without CRUD. I'll try to rename it quickly.

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:14]: If method is closer to JAX-RS which is what we're using I'm fine with it

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:12]: Maybe just `PanacheEntityResource`?

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (FroMage) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:14]: I'm only mentioning it because we started with “controller” and we got forced to adhere to JAX-RS's (mistaken) “resource”. So if we need to map all the way to JAX-RS probably it should be `@MethodProperties`?

BTW, that link you posted is non-normative. Reference for REST is https://www.ics.uci.edu/~fielding/pubs/dissertation/top.htm (and doesn't mention operations at all. I couldn't find at a glance how it refers to the operations you can do on the resources) but seems to call them “operations”.

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (FroMage) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:12]: I guess, yes.

Comment (FroMage) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:12]: Probably the same simplification could be done to the repository base name?

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:14]: Yeah I'll just rename to `@MethodProperties` to be inline with JAX-RS. I'm renaming the interface anyway, so will do it at the same time.

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:12]: Yes of course. Will do `PanacheEntityResource` and `PanacheRepositoryResource` for the ones defined in the Hibernate extension. And `RestDataResource` for that one defined in the common module that is extended by the first two.

Comment (FroMage) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java:12]: :+1:

Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java]: Done

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/CustomPathItemsResource.java]: Done

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/OperationProperties.java
Comment (gytis) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/OperationProperties.java]: Done

File: extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java
Comment (gytis) [reply extensions/panache/rest-data-panache/runtime/src/main/java/io/quarkus/rest/data/panache/ResourceProperties.java:16]: Done

**Comment by FroMage** [comment]
OK, so we're only missing the docs, right?

**Comment by gytis** [comment]
I've just cherry-picked the docs commit from this PR https://github.com/quarkusio/quarkus/pull/9254 so it would be easier for you to review and merge.

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:19]:
```suggestion
== Setting up REST Data with Panache
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:24]:
```suggestion
** Hibernate ORM REST Data with Panache extension (`quarkus-hibernate-orm-rest-data-panache`)
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:26]:
```suggestion
** One of the RESTEasy JSON serialization extensions (`quarkus-resteasy-jackson` or `quarkus-resteasy-jsonb`)
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:51]:
```suggestion
REST Data with Panache generates JAX-RS resources based on the interfaces available in your application.
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:57]:
```suggestion
If your application has an entity (e.g. `Person`) that extends either `PanacheEntity` or `PanacheEntityBase` class, you could instruct REST Data with Panache to generate its JAX-RS resource with the following interface:
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:67]:
```suggestion
If your application has a simple entity (e.g. `Person`) and a repository (e.g. `PersonRepository`) that implements either `PanacheRepository` or `PanacheRepositoryBase` interface, you could instruct REST Data with Panache to generate its JAX-RS resource with the following interface:
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:80]:
```suggestion
If you have defined one of the `PeopleResource` interfaces mentioned above, REST Data with Panache will generate a JAX-RS resource similar to this:
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:127]:
```suggestion
REST Data with Panache provides a `@ResourceProperties` and `@MethodProperties` annotations that can be used to customize certain features of the resource.
```

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (FroMage) [inline docs/src/main/asciidoc/rest-data-panache.adoc:157]:
```suggestion
As mentioned above REST Data with Panache supports the `application/json` and `application/hal+json` response content types.
```

**Review by FroMage — CHANGES_REQUESTED** [review]
OK, sorry I didn't realise we hadn't merged the docs PR yet. Thanks for bringing the commit over.

File: docs/src/main/asciidoc/rest-data-panache.adoc
Comment (gytis) [reply docs/src/main/asciidoc/rest-data-panache.adoc:19]: Sorry I moved it over quick, missed some obvious stuff.

**Comment by gytis** [comment]
Thanks for the review @FroMage. I've committed and squashed your suggestions.

**Comment by gytis** [comment]
OpenID test failed the CI. Shouldn't be related to this PR.

**Comment by geoand** [comment]
> OpenID test failed the CI. Shouldn't be related to this PR.

Yeah, it's a known flaky test

**Review by FroMage — APPROVED** [review]
Great!

**Comment by FroMage** [comment]
I've restarted the CI tests

**Comment by gytis** [comment]
"Native Tests - Data3" is the one that runs native tests for this extension. It passed our tests and later failed the `Reactive Pg Client`. It's not related with this change

**Comment by geoand** [comment]
> "Native Tests - Data3" is the one that runs native tests for this extension. It passed our tests and later failed the `Reactive Pg Client`. It's not related with this change

Yeah, thats the reason we added `--fail-at-end` :)

**Pull request merged**
